File: frontend/lib/services/auth_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../config/app_config.dart';
import '../models/user_model.dart';
import 'api_service.dart';

class AuthService {
  final ApiService _apiService = ApiService();

  // Register user
  Future<Map<String, dynamic>> register({
    required String name,
    required String email,
    required String password,
    required String phone,
  }) async {
    try {
      final response = await _apiService.post(
        '/auth/register',
        {
          'name': name,
          'email': email,
          'password': password,
          'phone': phone,
        },
        includeAuth: false,
      );

      if (response['success'] == true) {
        await _saveToken(response['token']);
        await _saveUser(response['data']);
        return {
          'success': true,
          'user': User.fromJson(response['data']),
          'token': response['token'],
        };
      }

      throw Exception(response['error'] ?? 'Registration failed');
    } catch (e) {
      throw Exception('Registration error: $e');
    }
  }

  // Login user
  Future<Map<String, dynamic>> login({
    required String email,
    required String password,
  }) async {
    try {
      final response = await _apiService.post(
        '/auth/login',
        {
          'email': email,
          'password': password,
        },
        includeAuth: false,
      );

      if (response['success'] == true) {
        await _saveToken(response['token']);
        await _saveUser(response['data']);
        return {
          'success': true,
          'user': User.fromJson(response['data']),
          'token': response['token'],
        };
      }

      throw Exception(response['error'] ?? 'Login failed');
    } catch (e) {
      throw Exception('Login error: $e');
    }
  }

  // Get current user
  Future<User> getCurrentUser() async {
    try {
      final response = await _apiService.get('/auth/me');
      if (response['success'] == true) {
        await _saveUser(response['data']);
        return User.fromJson(response['data']);
      }
      throw Exception('Failed to get user');
    } catch (e) {
      throw Exception('Get user error: $e');
    }
  }

  // Update password
  Future<void> updatePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    try {
      final response = await _apiService.put(
        '/auth/updatepassword',
        {
          'currentPassword': currentPassword,
          'newPassword': newPassword,
        },
      );

      if (response['success'] != true) {
        throw Exception(response['error'] ?? 'Password update failed');
      }
    } catch (e) {
      throw Exception('Password update error: $e');
    }
  }

  // Logout
  Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(AppConfig.tokenKey);
    await prefs.remove(AppConfig.userKey);
  }

  // Check if user is logged in
  Future<bool> isLoggedIn() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString(AppConfig.tokenKey);
    return token != null && token.isNotEmpty;
  }

  // Get saved user
  Future<User?> getSavedUser() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final userJson = prefs.getString(AppConfig.userKey);
      if (userJson != null) {
        return User.fromJson(json.decode(userJson));
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  // Save token
  Future<void> _saveToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(AppConfig.tokenKey, token);
  }

  // Save user
  Future<void> _saveUser(Map<String, dynamic> userData) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(AppConfig.userKey, json.encode(userData));
  }
}

File: frontend/src/services/authService.js
/**
 * authService.js - Authentication API Service
 * 
 * This file contains all API calls related to authentication:
 * - Register new user
 * - Login user
 * - Get current user profile
 * - Update password
 * 
 * Each function:
 * 1. Makes HTTP request to backend
 * 2. Returns response data
 * 3. Handles errors
 * 
 * Used by: AuthContext, Login/Register pages
 */

import api from './api'

/**
 * Register a new user
 * @param {Object} userData - { name, email, password, phone }
 * @returns {Promise} Response with token and user data
 */
export const register = async (userData) => {
  const response = await api.post('/auth/register', userData)
  return response.data
}

/**
 * Login user
 * @param {Object} credentials - { email, password }
 * @returns {Promise} Response with token and user data
 */
export const login = async (credentials) => {
  const response = await api.post('/auth/login', credentials)
  return response.data
}

/**
 * Get current user profile
 * @returns {Promise} User data
 */
export const getCurrentUser = async () => {
  const response = await api.get('/auth/me')
  return response.data
}

/**
 * Update user password
 * @param {Object} passwordData - { currentPassword, newPassword }
 * @returns {Promise} Success message
 */
export const updatePassword = async (passwordData) => {
  const response = await api.put('/auth/updatepassword', passwordData)
  return response.data
}
